packages/compartment-mapper/src/assemble.js

```diff
@@ -2,15 +2,110 @@
 /// <reference types="ses" />
 
 import { resolve } from './node-module-specifier.js';
-import { mapParsers } from './parse.js';
+import { parseExtension } from './extension.js';
 
-const { entries, freeze } = Object;
+const { entries, fromEntries, freeze } = Object;
 
 const defaultCompartment = Compartment;
 
 // q, as in quote, for strings in error messages.
 const q = JSON.stringify;
 
+/**
+ * @template {string | number} Key
+ * @template Value
+ * @param {Record<Key, Value>} object
+ * @param {Key} key
+ * @returns {boolean}
+ */
+const has = (object, key) => Object.prototype.hasOwnProperty.call(object, key);
+
+/**
+ * @param {Record<string, string>} languageForExtension - maps a file extension
+ * to the corresponding language.
+ * @param {Record<string, string>} languageForModuleSpecifier - In a rare case,
+ * the type of a module is implied by package.json and should not be inferred
+ * from its extension.
+ * @param {Record<string, ParseFn>} parserForLanguage
+ * @param {ModuleTransforms} transforms
+ * @returns {ParseFn}
+ */
+export const makeExtensionParser = (
+  languageForExtension,
+  languageForModuleSpecifier,
+  parserForLanguage,
+  transforms,
+) => {
+  return async (bytes, specifier, location, packageLocation) => {
+    let language;
+    if (
+      Object(languageForModuleSpecifier) === languageForModuleSpecifier &&
```

Comment:
Here you're ignoring `languageForModuleSpecifier` for any primitive value. Tracing through the callers, it looks like this is actually optional, so it should be typed that way. Ideally, an optional parameter should also use optional notation, but that's weird if the parameters after it are not optional.

Then, the above test should either be narrower `languageForModuleSpecifier !== undefined` or wider `typeof languageForModuleSpecifier === 'object'`. Otherwise, if someone passes a string, you'll just silently ignore it rather than giving them back a useful error. Also, think about what you'd want to specify.